src/sage/rings/polynomial/polynomial_element.pyx

```diff
@@ -2168,162 +2168,151 @@ cdef class Polynomial(CommutativePolynomial):
-            # now self has only roots of degree ``degree``.
-            # for now, we only implement the Cantor-Zassenhaus split
-            k = self.degree() // degree
-            if k == 1:
-                try:
-                    return self.roots(ring, multiplicities=False)[0] # is there something better to do here?
-                except IndexError:
-                    raise ValueError("no roots F %s" % self)
-            if q % 2 == 0:
-                while True:
-                    T = R.random_element(2*degree-1)
-                    if T == 0:
-                        continue
-                    T = T.monic()
-                    C = T
-                    for i in range(degree-1):
-                        C = T + pow(C,q,self)
-                    h = self.gcd(C)
-                    hd = h.degree()
-                    if hd != 0 and hd != self.degree():
-                        if 2*hd <= self.degree():
-                            return h.any_root(ring, -degree, True)
-                        else:
-                            return (self//h).any_root(ring, -degree, True)
-            else:
-                while True:
-                    T = R.random_element(2*degree-1)
-                    if T == 0:
-                        continue
+                return -f[0] / f[1]
+            return ring(-f[0] / f[1])
+
+        def _any_root_squarefree(f, ring=None):
+            # Compute the distinct degree factorisation for degree one, 
+            # the output will be the product of #all degree one 
+            # irreducible polynomials dividing f
+            q = f.base_ring().order() # p^k
+            R, X = f.parent().objgen()
+            w = pow(X, q, f) - X
+            g = f.gcd(w)
+
+            # When the degree is one, we have found our
+            # root. If the degree is zero, there are no roots
+            if g.degree().is_one():
+                return _linear_root(g, ring=ring)
+            elif g.degree().is_zero():
+                raise ValueError(f"no root can be computed for {f}")
+
+            # Otherwise, we can find a root from the CZ splitting of g
+            return _cantor_zassenhaus_split_root(g, 1, ring=ring)
+
+        def _cantor_zassenhaus_split_root(f, degree, ring=None):
+            R = f.parent()
+            q = f.base_ring().order()
+            # Need to handle odd and even charcteristic separately
+            if q % 2 != 0:
+                # We expect to succeed with 1/2 probability, so if we
+                # try 1000 times and fail, there's a bug somewhere.
+                for _ in range(1000):
+                    T = R.random_element(2*degree + 1)
+                    if T.is_zero(): continue # skip T = 0
                     T = T.monic()
-                    h = self.gcd(pow(T, Integer((q**degree-1)/2), self)-1)
+
+                    # Compute the gcd. 50% chance this is a non-trivial factor of f
```

Comment:
It's actually *at least* 50%: The probability of splitting is lower bounded by $1-(1/2)^{\deg(f)-1}$.